libraries/aws/shadow/src/shadow.c

```diff
@@ -0,0 +1,550 @@
+                                      uint8_t thingNameLength,
+                                      char * pTopicBuffer,
+                                      uint16_t bufferSize,
+                                      uint16_t * pOutLength )
+{
+    uint16_t offset = 0U, generatedTopicStringLength = 0U;
+    ShadowStatus_t shadowStatus = SHADOW_STATUS_SUCCESS;
+
+    /* Lookup table for Shadow operation string. */
+    static const char * const pTopicString[ SHADOW_TOPIC_STRING_TYPE_MAX_NUM ] =
+    {
+        SHADOW_TOPIC_OPERATION_STRING_GET,
+        SHADOW_TOPIC_OPERATION_STRING_GET_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_STRING_GET_REJECTED,
+        SHADOW_TOPIC_OPERATION_STRING_DELETE,
+        SHADOW_TOPIC_OPERATION_STRING_DELETE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_STRING_DELETE_REJECTED,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_REJECTED,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_DOCUMENTS,
+        SHADOW_TOPIC_OPERATION_STRING_UPDATE_DELTA
+    };
+
+    /* Lookup table for Shadow operation string length. */
+    static const uint16_t pTopicStringLength[ SHADOW_TOPIC_STRING_TYPE_MAX_NUM ] =
+    {
+        SHADOW_TOPIC_OPERATION_LENGTH_GET,
+        SHADOW_TOPIC_OPERATION_LENGTH_GET_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_GET_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_DELETE,
+        SHADOW_TOPIC_OPERATION_LENGTH_DELETE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_DELETE_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_ACCEPTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_REJECTED,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_DOCUMENTS,
+        SHADOW_TOPIC_OPERATION_LENGTH_UPDATE_DELTA
+    };
+
+    if( ( pTopicBuffer == NULL ) ||
+        ( pThingName == NULL ) ||
+        ( thingNameLength == 0U ) ||
+        ( topicType >= SHADOW_TOPIC_STRING_TYPE_MAX_NUM ) ||
+        ( pOutLength == NULL ) )
+    {
+        shadowStatus = SHADOW_STATUS_BAD_PARAMETER;
+        LogError( ( "Invalid input parameters" ) );
+    }
+    else
+    {
+        generatedTopicStringLength = SHADOW_TOPIC_PREFIX_LENGTH +       /* Prefix ("$aws/things/"). */
+                                     thingNameLength +                  /* Thing name. */
+                                     pTopicStringLength[ topicType ];   /* Shadow operation. */
+
+        if( bufferSize < generatedTopicStringLength )
+        {
+            shadowStatus = SHADOW_STATUS_BUFFER_TOO_SMALL;
+            LogError( ( "Input bufferSize too small, bufferSize %d, required ", bufferSize, generatedTopicStringLength) );
+        }
+        else
+        {
+            /* Copy the Shadow topic prefix into the topic buffer. */
+            ( void ) memcpy( ( void * ) pTopicBuffer,
+                             ( const void * ) SHADOW_TOPIC_PREFIX_STRING,
```

Comment:
it's because [memcpy
void * memcpy ( void * destination, const void * source, size_t num );](http://www.cplusplus.com/reference/cstring/memcpy/)

has void * and const void *, and const char * will raise flag for implicit casting